spec.html

```diff
@@ -339,35 +438,37 @@ <h1>InnerModuleEvaluation( _module_, _stack_, _index_ )</h1>
-          1. <del>Otherwise return _module_.[[EvaluationError]].</del>
+          1. If _module_.[[EvaluationError]] is *undefined*, return _index_.
+          1. Otherwise return _module_.[[EvaluationError]].
         1. If _module_.[[Status]] is `"evaluating"`, return _index_.
         1. Assert: _module_.[[Status]] is `"instantiated"`.
         1. Set _module_.[[Status]] to `"evaluating"`.
         1. Set _module_.[[DFSIndex]] to _index_.
         1. Set _module_.[[DFSAncestorIndex]] to _index_.
-        1. <ins>Let _evalCapability_ be ! NewPromiseCapability(%Promise%).</ins>
-        1. <ins>Set _module_.[[ExecPromise]] to _evalCapability_.[[Promise]].
         1. Set _index_ to _index_ + 1.
         1. Append _module_ to _stack_.
-        1. <ins>Let _dependencyExecPromises_ be an empty List.</ins>
+        1. <ins>Let _asyncDependencies_ be an empty List.</ins>
         1. For each String _required_ that is an element of _module_.[[RequestedModules]], do
           1. Let _requiredModule_ be ! HostResolveImportedModule(_module_, _required_).
           1. NOTE: Instantiate must be completed successfully prior to invoking this method, so every requested module is guaranteed to resolve successfully.
           1. Set _index_ to ? InnerModuleEvaluation(_requiredModule_, _stack_, _index_).
           1. Assert: _requiredModule_.[[Status]] is either `"evaluating"` or `"evaluated"`.
           1. Assert: _requiredModule_.[[Status]] is `"evaluating"` if and only if _requiredModule_ is in _stack_.
-          1. <ins>If _requiredModule_.[[Status]] is `"evaluated"`, then</ins>
-            1. <ins>Add _requiredModule_.[[ExecPromise]] to the list _dependencyExecPromises_.</ins>
+          1. <ins>If _requiredModule_.[[Status]] is `"evaluated"`, and _requiredModule_.[[Async]] is *true*, then</ins>
+            1. <ins>Assert: _module_.[[Async]] is *true*.</ins>
+            1. <ins>Append _requiredModule_.[[EvaluationPromise]] to _asyncDependencies_.</ins>
           1. If _requiredModule_.[[Status]] is `"evaluating"`, then
-            1. Assert: _requiredModule_ is a Source Text Module Record.
+            1. Assert: _requiredModule_ is a Cyclic Module Record.
             1. Set _module_.[[DFSAncestorIndex]] to min(_module_.[[DFSAncestorIndex]], _requiredModule_.[[DFSAncestorIndex]]).
-        1. <del>Perform ? ModuleExecution(_module_).</del>
-        1. <ins>Perform ! ExecuteModuleWhenImportsReady(_module_, _dependencyExecPromises_, _evalCapability_).</ins>
+        1. <ins>If _module_.[[Async]] is *false*, then</ins>
+          1. Perform ? _module_.ExecuteModule()
+        1. <ins>Otherwise,</ins>
+          1. <ins>Let _capability_ be ! NewPromiseCapability(%Promise%).</ins>
+          1. <ins>Set _module_.[[EvaluationPromise]] to _capability_.[[Promise]].</ins>
+          1. <ins>Perform ! ExecuteModuleWhenImportsReady(_module_, _asyncDependencies_, _capability_).</ins>
```

Comment:
Consider a graph A imports B and C, and B imports D and E in that order:

```
A -> B  -> D
         B  -> E
A -> C
```

The execution of the above graph when importing A is DEBCA.

Now if all of these modules are sync except for D, that means that we are entering this `ExecuteModuleWhenImportsReady` for the entire graph of E, B, C, A execution, just for D being async.

What execution guarantees do we provide for the execution of the sync sequence EBCA here? Would there be microtasks or anything of that sort running there?

Reply:
Basically what I'm saying is, surely BCA could still execute completely sync here even despite B having an async dependency?

Reply:
I think C and E would both execute synchronously, without any microtasks involved. The only Promise chains involved would be basically, D.then(() => B).then(() => A). B and A wait on D, so they can't start synchronously.

Reply:
But wouldn't it be more like ` D.then(() => Promise.all(B, C)).then(() => A)`? I guess the most concerning here is the C then A part being synchronous, since this means if I just have A->C and I add an import to B above the import to C in A, I'm now losing the execution order of C happening just before A synchronously? Or am I misdiagnosing this?

Reply:
C doesn't participate in the Promise.all, since it's synchronous.

Reply:
Ahh, so C is being executed early in parallel with B, because async dependencies don't block.

So if we had console.log statements in each of the execution bodies, with the async module of D having a console.log('startD') and console.log('endD'), then would the output be the following:

```
start D
E
C
end D
B
A
```

I wonder if the conceptual model might be simpler and more deterministic if we treat sync imports as hoisted always? So that we would instead get:

```
E
C
start D
end D
B
A
```

That might be useful in build tools as well as then it's easier to optimize based on knowing the sync execution guarantees.

Reply:
Or if not hoisted, to at least know there is an event loop yield before D starts execution? Or is that already in place?

Reply:
Your first summary is the ordering I was expecting. What would be the motivation for the other ordering? It seems more counterintuitive to me.

Reply:
I can get behind the first ordering as long as it is fully well-defined - that is that synchronous first part of the async function runs in normal sync execution order without invoking any microtasks before its execution. So if there was a dependency F that was synchronous before D, that FD would execute in sequence completely synchronously without microtasks or yielding. As long as that is ok, this sounds incredibly well defined to me.

Reply:
(where D above is the synchronous part of D before its first "await" statement)

Reply:
Yes, that's what this spec text (aims to) describe.

Reply:
If that is the case, then this all sounds excellent to me!